library/core/src/intrinsics/simd.rs

```diff
@@ -0,0 +1,473 @@
+    ///
+    /// `T` must be a vector.
+    ///
+    /// `U` must be a const array of `i32`s.
+    ///
+    /// `V` must be a vector with the same element type as `T` and the same length as `U`.
+    ///
+    /// Concatenates `x` and `y`, then returns a new vector such that each element is selected from
+    /// the concatenation by the matching index in `idx`.
+    pub fn simd_shuffle<T, U, V>(x: T, y: T, idx: U) -> V;
+
+    /// Read a vector of pointers.
+    ///
+    /// `T` must be a vector.
+    ///
+    /// `U` must be a vector of pointers to the element type of `T`, with the same length as `T`.
+    ///
+    /// `V` must be a vector of integers with the same length as `T` (but any element size).
+    ///
+    /// `idx` must be a constant: either naming a constant item, or an inline
+    /// `const {}` expression.
+    ///
+    /// For each pointer in `ptr`, if the corresponding value in `mask` is `!0`, read the pointer.
+    /// Otherwise if the corresponding value in `mask` is `0`, return the corresponding value from
+    /// `val`.
+    ///
+    /// # Safety
+    /// Unmasked values in `T` must be readable as if by `<ptr>::read` (e.g. aligned to the element
+    /// type).
+    ///
+    /// `mask` must only contain `0` or `!0` values.
+    pub fn simd_gather<T, U, V>(val: T, ptr: U, mask: V) -> T;
+
+    /// Write to a vector of pointers.
+    ///
+    /// `T` must be a vector.
+    ///
+    /// `U` must be a vector of pointers to the element type of `T`, with the same length as `T`.
+    ///
+    /// `V` must be a vector of integers with the same length as `T` (but any element size).
+    ///
+    /// For each pointer in `ptr`, if the corresponding value in `mask` is `!0`, write the
+    /// corresponding value in `val` to the pointer.
+    /// Otherwise if the corresponding value in `mask` is `0`, do nothing.
+    ///
+    /// # Safety
+    /// Unmasked values in `T` must be writeable as if by `<ptr>::write` (e.g. aligned to the element
+    /// type).
+    ///
+    /// `mask` must only contain `0` or `!0` values.
+    pub fn simd_scatter<T, U, V>(val: T, ptr: U, mask: V);
+
+    /// Read a vector of pointers.
+    ///
+    /// `T` must be a vector.
+    ///
+    /// `U` must be a vector of pointers to the element type of `T`, with the same length as `T`.
+    ///
+    /// `V` must be a vector of integers with the same length as `T` (but any element size).
+    ///
+    /// For each element, if the corresponding value in `mask` is `!0`, read the corresponding
+    /// pointer from `ptr`.
+    /// Otherwise if the corresponding value in `mask` is `0`, return the corresponding value from
+    /// `val`.
+    ///
+    /// # Safety
+    /// Unmasked values in `T` must be readable as if by `<ptr>::read` (e.g. aligned to the element
+    /// type).
+    ///
+    /// `mask` must only contain `0` or `!0` values.
+    #[cfg(not(bootstrap))]
+    pub fn simd_masked_load<V, U, T>(mask: V, ptr: U, val: T) -> T;
```

Comment:
In implementing this, I am confused. Isn't this equivalent to `simd_gather`?

Reply:
gather accepts a vector of N pointers, where each element will be loaded from its corresponding pointer.

masked_load accepts a single pointer and all elements of the resulting vector (when unmasked) are loadded from a constant offset from that pointer. i.e the first element will be loaded from `ptr`, second from `ptr.offset(1)`, and so on.

Reply:
Then the documentation here is wrong. For masked_load it says
```
/// `U` must be a vector of pointers to the element type of `T`, with the same length as `T`.
```

Reply:
Yeah, looks like we missed this. I'll create a follow up PR soon

Reply:
https://github.com/rust-lang/rust/pull/119203